### ISSUE TITLE
targeting variables and functions created in JSXGraph-block

### ISSUE THREAD

**Issue by Famondir** [maths/moodle-qtype_stack#942]
Hi all,
I'ld started working with JSXGraph in Stack-questions. I realy like the possibility to let the system manage multiple JSXGraphs on one quiz page using the [[jsxgraph]] block. The functions for restoring the state of an interactive graph are also pretty handy.

I want to make an interactive graph where students can add points by clicking on a button outside of the JSX panel. Using JSXGraph without the [[jsxgraph]] block makes it easy to access all the variables and functions. Is there a way, to access the variables and functions in the element created by the [[jsxgraph]] block as well?

E.g. by giving the block an identifier [[jsxgraph name="fu"]] and referencing the variables and methods like fu would be an object?: fu.function1() or fu.setVariable()?

What process is happening in the back when someone is using the [[jsxgraph]] block?

Sincerely Simon

P.S.: I found the possibility to create buttons in JSXGraph. Those work just fine. I could also move those with JS under their parent div. But I'm curious about what happens behind the scenes with [[jsxgraph]] block and if one could interact with the objects created there.

**Comment by aharjula** [comment]
No.

The reason is security related. While in the current system, that security blocks access to the bits inside the block, in the future, we will make the security system stricter. So that the block will cease to have any access to anything that we do not specifically provide outside itself, in the future, the security will be enforced by a sandboxed iframe.

In the current (and I promise) in the future system, one can simply bind multiple `[[jsxgraph]]`-blocks to the same input (e.g. a string input not used for the question and storing a JSON object) and use that to share state, not functions though. This input-focused approach has a benefit that variables do not, as it allows one to attach change-listeners to those inputs and thus react to the values changing (one just needs to trigger that change event).

The future system relies on [message passing](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage), and we will provide binding logic that hides that message passing and provides cloned and synchronised input elements within the `[[jsxgraph]]`-block. If one really wants to share state between such blocks but does not want to do it through inputs, one could pass such messages, although in the current system, that makes little sense.


As to what exactly happens in the `[[jsxgraph]]`-block currently is that the content of the block gets extracted from the document and injected back as a JavaScript fragment with [`"use strict";`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) added with those identifiers for the div and the referenced inputs. The injection happens in such a way that it almost looks like a JavaScript module, thus blocking access into it. In the current world, the parent document is still currently accessible and thus, those `stack_jxg` binding functions can directly attach to it. In the future, those will bind to cloned inputs provided by the new library called [`stack_js`](https://github.com/maths/moodle-qtype_stack/blob/7bba8e99503ab80c93be8f096396ec5a5222debd/corsscripts/stackjsiframe.js#L153-L247) which will be the gatekeeper between stuff happening inside blocks and stuff happening on the VLE side.

That gatekeeper has two roles, firstly, it removes certain security risks related to running arbitrary code inside a VLE where that code could hijack teacher rights, and secondly, it provides a simple API for the block contents to act with whatever VLE is hosting that block.

Basically, in the future, you may not use JavaScript placed inside any of these blocks to affect the VLE around them. This is a limitation, but if we do not add that limitation, you won't be doing anything, as people will disable those blocks. This is not because a random author is always aiming to attack something. It is because materials get shared and used by people that do not necessarily know what they are using, thus we need to make it so that the material cannot "do evil".
